internal/catalogmetadata/client/client_test.go

```diff
@@ -60,59 +61,64 @@ func TestClient(t *testing.T) {
 										"name": "fake1.v9.9.9"
 									}
 								]
-							}`),
-						},
-					})
+							}`)...)
 
-					return objs, nil
+					return objs, nil, catalogContentMap
 				},
 				wantErr: `bundle "fake1.v9.9.9" not found in catalog "catalog-1" (package "fake1", channel "channel-with-missing-bundle")`,
 			},
+			{
+				name: "invalid meta",
+				fakeCatalog: func() ([]client.Object, []*catalogmetadata.Bundle, map[string][]byte) {
+					objs, _, catalogContentMap := defaultFakeCatalog()
+
+					catalogContentMap["catalog-1"] = append(catalogContentMap["catalog-1"], []byte(`{"schema": "olm.bundle", "name":123123123}`)...)
+
+					return objs, nil, catalogContentMap
+				},
+				wantErr: "error processing response: error was provided to the WalkMetasReaderFunc: expected value for key \"name\" to be a string, got %!t(float64=1.23123123e+08): 1.23123123e+08",
+			},
 			{
 				name: "invalid bundle",
-				fakeCatalog: func() ([]client.Object, []*catalogmetadata.Bundle) {
-					objs, _ := defaultFakeCatalog()
+				fakeCatalog: func() ([]client.Object, []*catalogmetadata.Bundle, map[string][]byte) {
+					objs, _, catalogContentMap := defaultFakeCatalog()
 
-					objs = append(objs, &catalogd.CatalogMetadata{
-						ObjectMeta: metav1.ObjectMeta{
-							Name:   "catalog-1-broken-bundle",
-							Labels: map[string]string{"schema": declcfg.SchemaBundle, "catalog": "catalog-1"},
-						},
-						Spec: catalogd.CatalogMetadataSpec{
-							Content: json.RawMessage(`{"name":123123123}`),
-						},
-					})
+					catalogContentMap["catalog-1"] = append(catalogContentMap["catalog-1"],
+						[]byte(`{"schema": "olm.bundle", "name":"foo", "package":"bar", "image":123123123}`)...)
 
-					return objs, nil
+					return objs, nil, catalogContentMap
 				},
-				wantErr: "error unmarshalling catalog metadata: json: cannot unmarshal number into Go struct field Bundle.name of type string",
+				wantErr: "error processing response: error unmarshalling bundle from catalog metadata: json: cannot unmarshal number into Go struct field Bundle.image of type string",
```

Comment:
Global comment: errors are opaque values in Go. If a library author wants you to be able to inspect a particular error, they will export something and you will be able to use `errors.As` on it. If not, there is nothing to be gained from testing the specific strings emitted as part of the error - all it does is create extremely brittle unit tests that need to change every time the implementation does.

Reply:
Done in https://github.com/operator-framework/operator-controller/pull/411/commits/1cbe6157a213246232930aa3ce5f7259e6574c61

Reply:
>library maintainer you are very unlikely to spot an issue where refactored test or refactored implementation starts hitting a different code path. This happened even in https://github.com/operator-framework/operator-controller/pull/411#discussion_r1328477900.

I think the issue in this PR was spotted due to investigating the test coverage and realizing that a particular path wasn't covered and not necessarily because of the error strings. 

That being said, I don't have a strong preference here (although I do lean more towards creating less brittle unit tests) and am happy to do whatever folks prefer.

Reply:
> I think the issue in this PR was spotted due to investigating the test coverage and realizing that a particular path wasn't covered and not necessarily because of the error strings.

No it was the fact that the expected error string changed made me wonder whether it was correct change or not. I used code coverage to confirm and illustrate the point.

This is something I catch fairly often this way and I'm not aware of the alternative.

This is not the first time I see the same opinions on error test (another example I could quickly find is this [SO answer](https://stackoverflow.com/a/42036847) too). However I've never heard of an alternative which prevents this and avoids exposing errors as an API (we do not need a custom error type in most of the cases).

This [blog post](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) also says "Never inspect the output of error.Error", however it also says:

> I know that sometimes this isn’t possible, and as someone pointed out on twitter, this advice doesn’t apply to writing tests. Never the less, comparing the string form of an error is, in my opinion, a code smell, and you should try to avoid it.

I realise that checking error string is not 100% guarantee from a regression + different code paths might lead to the same error string, but I think, in tests, it is still better than just "err != nil` check.

Reply:
>No it was the fact that the expected error string changed made me wonder whether it was correct change or not. I used code coverage to confirm and illustrate the point.

Fair enough. I'm not strongly opinionated on this matter so I'll go ahead and revert it to checking the strings. If there is further discussion desired on this I'd vote that we push this to a separate discussion around testing standards/expectations and keep it from blocking this PR.

Reply:
Reverted the change to `bool` in the latest push. Resolving this - if there is desire for further discussion please create a new issue/discussion about testing standards/expectations, thank you!

Reply:
I've encountered this same conundrum as well. Here was my answer:
- https://github.com/operator-framework/operator-registry/blob/eaebcf63f94dabdabc3a25ceeeedcb26aaa11d79/alpha/model/model_test.go#L84-L109
- https://github.com/operator-framework/operator-registry/blob/eaebcf63f94dabdabc3a25ceeeedcb26aaa11d79/alpha/model/model_test.go#L174-L198